src/win/util.c

```diff
@@ -1663,6 +1673,126 @@ int uv_os_unsetenv(const char* name) {
 }
 
 
+static int WSAAPI uv__gethostnamew_nt60(PWSTR name, int name_len) {
+  int result_len;
+  int error_code = NO_ERROR;
+
+  /* WSALookupService stuff
+   * Avoid dynamic memory allocation if possible */
+  CHAR local_buf[WSAQ_LOCAL_BUF_LEN];
+  DWORD dwlen = WSAQ_LOCAL_BUF_LEN;
+  WSAQUERYSETW* pwsaq;
+  /* hostname returned from WSALookupService stage */
+  WCHAR* result_name = NULL;
+  /* WSALookupService handle */
+  HANDLE hlookup;
+  /* Fallback to heap allocation if stack buffer is too small */
+  WSAQUERYSETW* heap_data = NULL;
+
+  /* check input */
+  if (name == NULL) {
+    error_code = WSAEFAULT;
+    goto cleanup;
+  }
+
+  /* 
+   * Stage 1: Check environment variable
+   * _CLUSTER_NETWORK_NAME_ len == ComputeName(NETBIOS) len.
+   * i.e 15 characters + null.
+   * It overrides the actual hostname, so application can
+   * work when network name and computer name are different 
+   */ 
+  result_len = GetEnvironmentVariableW(L"_CLUSTER_NETWORK_NAME_",
+                                       name,
+                                       name_len);
+  if (result_len != 0) {
+    if (result_len > name_len) {
+      error_code = WSAEFAULT;
+    }
+    goto cleanup;
+  }
+
+  /* Stage 2: Do normal lookup through WSALookupServiceLookup */
+  pwsaq = (WSAQUERYSETW*) local_buf;
+  memset(pwsaq, 0, sizeof(*pwsaq));
+  pwsaq->dwSize                  = sizeof(*pwsaq);
+  pwsaq->lpszServiceInstanceName = NULL;
+  pwsaq->lpServiceClassId        = &guid_host_name;
+  pwsaq->dwNameSpace             = NS_ALL;
+  pwsaq->lpafpProtocols          = &af_protocols[0];
+  pwsaq->dwNumberOfProtocols     = 2;
+
+  error_code = WSALookupServiceBeginW(pwsaq, LUP_RETURN_NAME, &hlookup);
+  if (error_code == NO_ERROR) {
+    /* Try stack allocation first */
+    error_code = WSALookupServiceNextW(hlookup, 0, &dwlen, pwsaq);
+    if (error_code == NO_ERROR) {
+      result_name = pwsaq->lpszServiceInstanceName;
+    } else {
+      error_code = WSAGetLastError();
+
+      if ((error_code == WSAEFAULT) && (dwlen > WSAQ_LOCAL_BUF_LEN)) {
```

Comment:
It seems that there is no need to test `dwlen > WSAQ_LOCAL_BUF_LEN` if result is `WSAEFAULT`?

Reply:
The reason I've added it in the first place was that it looked strange to me that WSAEFAULT were used as buffer size check and there can be other cases or hidden bugs. It turned out they've used WSAENIVAL instead.